### ISSUE TITLE
Emulator <-> API connection requires 4 API calls per frame.

### ISSUE THREAD

**Issue by OhmGeek** [georgeprice/n64all#17]
This is a slight bottleneck, as the emulator input plugin blocks while the TCP socket fetches the response, causing stuttering.

Instead of requesting 4 responses per image frame, instead request 1 response containing the entire memory buffer.

This will require a slight change in the JSON structure, such that:
```json
{
  "PLAYER_1_ANALOG": [0,0],
   ...

   "PLAYER_4_A_BTN: 0
}
```

We can then parse this more easily, using only function calls on a string instead of blocking the input thread.

**Comment by OliMac1** [comment]
New idea for reducing this bottleneck:

- UDP sent from the server to the plugin at set rate eg. 30 Hz.
- The plugin stores the relevant data and the response is fetched internally.
- To deal with potential atomicity issues I propose a system similar to swapping a framebuffer with one controller to be written to and one to be read from.

This stops the API being restful but will reduce latency and network traffic.

**Comment by OhmGeek** [comment]
That could definitely work - for now, I'm looking into running the constant polling in a separate thread, with controller still being stored in shared memory (and just being written). The bottleneck was because of the blocking nature, so that might help in the short term. But definitely switching to either UDP/using Kafka (and polling at 60 Hz as per Nyquist theorem might be suitable)

**Comment by OhmGeek** [comment]
See #20 for the multithreaded approach. Still working on testing it (need to sort out my linux again).

As part of the new Java server (#19), looking at making it fairly decoupled so we can accept user requests from REST, while throwing state outputs to either REST/UDP/TCP or even Apache Kafka (which might be useful if we want to have it dealing with multiple games in the future). 

If you want to start looking into the UDP based method that'll be good, as I think that'll be the best solution for now (Kafka is definitely a pipedream at the moment)

**Comment by OliMac1** [comment]
Do kafka streams support non Java consumers? The Plugin won't necessarily like the data fed to it?

**Comment by OhmGeek** [comment]
Yes, just depends on the serialiser/deserialiser: we would need to use a Json Serialise/Deserialisation based scheme, rather than bytes directly. But there is a C API for Kafka. UDP for now might be the best shout, since we only have one session, providing we fix the hostname based issue (so we could host the central server somewhere other than locally)
